presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveRedundantTableScanPredicate.java

```diff
@@ -0,0 +1,159 @@
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static java.util.Objects.requireNonNull;
+
+public class RemoveRedundantTableScanPredicate
+        implements Rule<FilterNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<FilterNode> PATTERN =
+            filter().with(source().matching(
+                    tableScan().capturedAs(TABLE_SCAN)));
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+
+    public RemoveRedundantTableScanPredicate(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.domainTranslator = new DomainTranslator(new LiteralEncoder(metadata));
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode filterNode, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        PlanNode rewritten = removeRedundantTableScanPredicate(
+                tableScan,
+                filterNode.getPredicate(),
+                context.getSession(),
+                context.getSymbolAllocator().getTypes(),
+                context.getIdAllocator());
+
+        if (arePlansSame(filterNode, tableScan, rewritten)) {
+            return Result.empty();
+        }
+
+        return Result.ofPlanNode(rewritten);
+    }
+
+    private PlanNode removeRedundantTableScanPredicate(
+            TableScanNode node,
+            Expression predicate,
+            Session session,
+            TypeProvider types,
+            PlanNodeIdAllocator idAllocator)
+    {
+        Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
+        Expression nonDeterministicPredicate = filterNonDeterministicConjuncts(predicate);
+
+        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+                metadata,
+                session,
+                deterministicPredicate,
+                types);
+
+        TupleDomain<ColumnHandle> predicateDomain = decomposedPredicate.getTupleDomain()
+                .transform(node.getAssignments()::get);
+
+        TupleDomain<ColumnHandle> unenforcedDomain;
+        if (predicateDomain.getDomains().isPresent()) {
+            Map<ColumnHandle, Domain> predicateColumnDomains = predicateDomain.getDomains().get();
+
+            // table scans with none domain should be converted to ValuesNode
+            checkState(node.getEnforcedConstraint().getDomains().isPresent());
+            Map<ColumnHandle, Domain> enforcedColumnDomains = node.getEnforcedConstraint().getDomains().get();
+
+            ImmutableMap.Builder<ColumnHandle, Domain> unenforcedColumnDomains = ImmutableMap.builder();
+            for (ColumnHandle columnHandle : predicateColumnDomains.keySet()) {
```

Comment:
Unless I'm misreading this, if the table scan domain for the column is `ALL`, this will do:

```
intersect(ALL, f) -> f
ALL.contains(f) -> true
```

Which will result in `f` being dropped.

I think what you want is `intersect(F, P) == P? => drop F` (where `F` is the filter, `P` is the table scan predicate). See below:


![IMG_7915](https://user-images.githubusercontent.com/14387/65807046-8c087880-e141-11e9-92ef-5e867894abfc.jpg)

Putting it all together:

```java
for (Map.Entry<ColumnHandle, Domain> entry : predicateColumnDomains.entrySet()) {
    ColumnHandle column = entry.getKey();
    Domain filter = entry.getValue();
    Domain columnPredicate = enforcedColumnDomains.getOrDefault(column, Domain.all(filter.getType()));

    if (!filter.intersect(columnPredicate).equals(columnPredicate)) {
        unenforcedColumnDomains.put(column, filter);
    }
}
```